app/src/main/java/org/dslul/openboard/inputmethod/latin/settings/UserDictionaryListFragment.java

```diff
@@ -0,0 +1,306 @@
+        // List of main language
+        final List<InputMethodSubtype> enabledMainSubtype = SubtypeSettingsKt.getEnabledSubtypes(prefs, true);
+        // List of enabled system language
+        final List<Locale> enabledSystemLocale = SubtypeSettingsKt.getSystemLocales();
+        // To combine lists and handle duplicates automatically
+        LinkedHashSet<String> userDictionaryLanguage = new LinkedHashSet<>();
+
+        // Add "For all languages"
+        userDictionaryLanguage.add("");
+
+        // Add the language from the user dictionary if a word is present
+        if (enabledUserDictionary != null) {
+            userDictionaryLanguage.addAll(enabledUserDictionary);
+        }
+
+        // Add the main language selected in the "Language and Layouts" setting
+        for (InputMethodSubtype mainSubtype : enabledMainSubtype) {
+            Locale mainLocale = LocaleUtils.constructLocaleFromString(mainSubtype.getLocale());
+            String mainLocaleString = mainLocale.toString();
+            // Special treatment for the known languages with _ZZ and _zz types
+            if (mainLocaleString.endsWith("_ZZ") || mainLocaleString.endsWith("_zz") || mainLocaleString.equals("zz")) {
+                mainLocaleString = LocaleUtils.getLocaleDisplayNameInSystemLocale(mainLocale, requireActivity()).toLowerCase();
+            }
+            userDictionaryLanguage.add(mainLocaleString);
+
+            // Add secondary language only if main language is selected and "Use system languages" setting is not enabled
+            for (Locale secondSubtype : Settings.getSecondaryLocales(prefs, String.valueOf(mainLocale))) {
+                Locale secondLocale = LocaleUtils.constructLocaleFromString(String.valueOf(secondSubtype));
+                String secondLocaleString = secondLocale.toString();
+                // Special treatment for the known languages with _ZZ and _zz types
+                if (secondLocaleString.endsWith("_ZZ") || secondLocaleString.endsWith("_zz") || secondLocaleString.equals("zz")) {
+                    secondLocaleString = LocaleUtils.getLocaleDisplayNameInSystemLocale(secondLocale, requireActivity()).toLowerCase();
+                }
+                if (!localeSystemOnly) {
+                    userDictionaryLanguage.add(secondLocaleString);
+                }
+            }
+        }
+
+        // Finally, add the enabled system languages
+        for (Locale subtype : enabledSystemLocale) {
+            Locale systemLocale = LocaleUtils.constructLocaleFromString(String.valueOf(subtype));
+            String systemLocaleString = systemLocale.toString();
+            // Special treatment for the known languages with _ZZ and _zz types
+            if (systemLocaleString.endsWith("_ZZ") || systemLocaleString.endsWith("_zz") || systemLocaleString.equals("zz")) {
+                systemLocaleString = LocaleUtils.getLocaleDisplayNameInSystemLocale(systemLocale, requireActivity()).toLowerCase();
+            }
```

Comment:
You use these lines several times. In such a case it's better move them to a separate function.

Reply:
Hmm, actually I don't understand the whole function.
You start with a (non-null) `Locale` get the string value (using the unusual and null-safe `String.valueOf`), and then create a locale from it again, only to convert it to a string again, this time using `toString`.
But if it's a `zz` locale, you use the lowercase display name instead of a locale string.

Reply:
Rectified with commit f950744